Add support for generating Bidi_Class tables

wezm opened the pull request:
This PR adds a `bidi-class` command. Sample output:

```rust
// DO NOT EDIT THIS FILE. IT WAS AUTOMATICALLY GENERATED BY:
//
//  ucd-generate bidi-class --rust-enum /home/wmoore/Downloads/ucd-12.1
//
// ucd-generate is available on crates.io.

#[derive(Clone, Copy, Debug, Hash, Eq, PartialEq)]
pub enum BidiClass {
  ArabicLetter, ArabicNumber, BoundaryNeutral, CommonSeparator,
  EuropeanNumber, EuropeanSeparator, EuropeanTerminator, FirstStrongIsolate,
  LeftToRight, LeftToRightEmbedding, LeftToRightIsolate, LeftToRightOverride,
  NonspacingMark, OtherNeutral, ParagraphSeparator, PopDirectionalFormat,
  PopDirectionalIsolate, RightToLeft, RightToLeftEmbedding,
  RightToLeftIsolate, RightToLeftOverride, SegmentSeparator, WhiteSpace,
}

pub const BIDI_CLASS: &'static [(u32, u32, BidiClass)] = &[
  (0, 8, BidiClass::BoundaryNeutral), (9, 9, BidiClass::SegmentSeparator),
  (10, 10, BidiClass::ParagraphSeparator),
  (11, 11, BidiClass::SegmentSeparator), (12, 12, BidiClass::WhiteSpace),
  (13, 13, BidiClass::ParagraphSeparator),
  (14, 27, BidiClass::BoundaryNeutral),
  ⋮
];
```

wezm committed:
Add support for generating Bidi_Class tables

BurntSushi commented on src/bidi_class.rs:
```diff
@@ -0,0 +1,145 @@
+    //
+    //  All code points not explicitly listed for Bidi_Class
+    //  have the value Left_To_Right (L).
+    #[rustfmt::skip]
```
Hmmm... Could you say more about this? I'd really rather we stuck to rustfmt unless there's a really good reason. In particular, for this code, it exceeds my normal 79 column limit.

wezm replied:
Basically my thinking was along these lines:

* The comment was copied directly from `DerivedBidiClass.txt`. I didn't reformat it so that future updates from the spec would show only what changed in a diff
* `default_class_assignments` was the code the represented the data contained in the comment/spec. I tried to format/group the data to match the layout in the spec to make it easier to eyeball that the two were the same. I figured this might also be useful if the need arose to update it in the future. I added the `rustfmt::skip` to avoid rustfmt messing with that formatting.

Anyway I've moved it to a constant because once you pointed it out I'm not really sure what I was thinking sticking it inline like that. Since the comment and data are tied to each other I wasn't sure this would work will inside `lookup_unassigned` so I put it at the top level.

BurntSushi commented on src/bidi_class.rs:
```diff
@@ -0,0 +1,145 @@
+    // For all other cases:
+    //
+    //  All code points not explicitly listed for Bidi_Class
+    //  have the value Left_To_Right (L).
```
Could you try to wrap this comment to 79 columns inclusive please?

BurntSushi commented on src/bidi_class.rs:
```diff
@@ -0,0 +1,145 @@
+    let left_to_right_name = bidi_class_name("L")?;
+    let boundary_neutral_name = bidi_class_name("BN")?;
+    for cp in 0..=0x10FFFF {
+        if !assigned.contains(&cp) {
```
I'd write it like this:

```rust
if assigned.contains(&cp) {
    continue;
}
if let Some(class) = ... {
    ...
}
```

I generally prefer this style because it flows a bit nicer and results in less nested code.

BurntSushi commented on src/bidi_class.rs:
```diff
@@ -0,0 +1,145 @@
+        (0x0001EF00, 0x0001EFFF, "R"),
+
+        (0x20A0, 0x20CF, "ET")
+    ];
```
Perhaps it would be better to move this to a constant? Or perhaps, define it as a constant inside of the `lookup_unassigned` helper function?

BurntSushi commented on src/bidi_class.rs:
```diff
@@ -0,0 +1,145 @@
+) -> Option<&'a str> {
+    defaults
+        .iter()
+        .find(|&&(start, end, _)| codepoint >= start && codepoint <= end)
```
For range tests like this, I'd rather the style `start <= codepoint && codepoint <= end`, which I think flows a bit better.

BurntSushi reviewed (commented):
Overall, this looks great! I think I just have a few mostly minor/stylistic comments. Nice work. :-)

wezm committed:
Address PR feedback

wezm commented:
Thanks for the quick review. I've addressed feedback. I'm not sure if you prefer fixes like to this to be applied and rebased or pushed as new commits. I've pushed the commit for now but happy to rebase it into the original commits if preferred.

BurntSushi reviewed (approved):
Awesome, thanks.

And yeah, it's a good point about replicating the same structure as the source of the comment. To be honest, I've usually tried not to copy Unicode spec stuff into comments like this and instead simply reference it, since it's an easily accessible resource.

With respect to rebasing... If you take a look at the commit logs of my projects, I do try to maintain a fairly clean linear history. It's for practical purposes (bisecting is easier) but also for just generally trying to make it easy to audit the history of a project. I generally fall in the camp of "the history is part of the presentation of the project" and generally don't care as much about development history.

As far as PRs go, if it's a pretty small PR like this one, then yeah, you can just amend the commits and force push. For a bigger PR, adding more commits will probably make review easier. Then we can clean things up before merging.
